### PR TITLE
Add basic dualization comparison to simple examples

### PR DESCRIPTION
**Pull request by blegat**
Alternative to https://github.com/jump-dev/JuMP.jl/pull/3408

Output is
```
Model Name         #PrimalVars #Cons #DualVars #DualCons
------------------------------------------------------------
Maximum cut             10      5       4       1
K-means                 21      29      28      22
Correlation             6       8       7       5
Minimum distortion      11      15      14      15
Theta                   15      7       6       1
Robust                  12      4       24      10
```

It's kind of a proof of concept, just counting the number of constraints is a bit deceiving, we should have one column for each constraint type.
I makes me think that it would be nice to have a nice way to print the inner model the same way as `show(::AbstractModel)`.
We're still throwing the user in a fight with a sword and a bow without a good way to see which one fits best.
If we had a way to have the same output as `show(::AbstractModel)` but for the inner optimizer, it's be great.
Could be something like
```julia
julia> show_inner_models(model)
A Bridge layer
Minimization problem with:
Variables: 2
`MOI.ScalarAffineFunction{Float64}`-in-`MOI.EqualTo{Float64}`: 10 constraints
...

A CachingOptimizer{SCS.OptimizerCache,SCS.Optimizer}
Minimization problem with:
Variables: 2
`MOI.VectorAffineFunction{Float64}`-in-`MOI.Zeros`: 10 constraints
...
```
and then
```julia
julia> show_inner_models(dualized_model)
A Bridge layer
Minimization problem with:
Variables: 2
`MOI.ScalarAffineFunction{Float64}`-in-`MOI.EqualTo{Float64}`: 10 constraints
...

A DualOptimizer
Minimization problem with:
Variables: 2
`MOI.VectorAffineFunction{Float64}`-in-`MOI.Zeros`: 10 constraints
...

A CachingOptimizer{SCS.OptimizerCache,SCS.Optimizer}
Minimization problem with:
Variables: 2
`MOI.VectorAffineFunction{Float64}`-in-`MOI.Zeros`: 10 constraints
...
```
To walk through layers, we would do something like
```julia
MOI.inner_optimizer(::MOI.ModelLike) = nothing
# We implement it for CachingOptimizer, LazyBridgeOptimizer and Dualization.DualOptimizer

function show_inner_models(model::MOI.ModelLike)
    _show_model(model) # Could do some try catch here in case the solver does not support the getters
    show_inner_models(MOI.inner_optimizer(model))
end
```
and `MOI.inner_optimizer` would also be useful for `JuMP.unsafe_backend`. It would also be a way for the user to inspect the list of layers.
Meta-solver could also implement `MOI.inner_optimizer` unless they have two inner optimizer in which case they will just not implement it.

### CODE REVIEW

File: docs/src/tutorials/conic/simple_examples.jl [hunk 1]
```diff
@@ -416,7 +435,58 @@ function example_robust_uncertainty_sets()
...
+
+function _num_variables(model::MOI.ModelLike)
+    return MOI.get(model, MOI.NumberOfVariables())
+end
+function _num_constraints(model::MOI.ModelLike)
+    return sum(
+        MOI.get(model, MOI.NumberOfConstraints{F,S}())
+        for (F, S) in MOI.get(model, MOI.ListOfConstraintTypesPresent())
+    )
+end
```
Comment (odow) [inline docs/src/tutorials/conic/simple_examples.jl:458]: This is too complicated for a "simple examples" tutorial

**Comment by odow** [comment]
I don't think we need this. The Dualization tutorial explains how to solve the primal and dual with SCS, and how to interpret the log to see which problem was smaller. I don't want to complicate the "simple examples" tutorial with this. And we definitely don't want to encourage general users to explore the inner workings of a JuMP model. "Try both" is okay.

**Comment by jd-foster** [comment]
This could be the basis for a nice blog post introducing the Dualization package, or adapted into its package documentation.

**Comment by odow** [comment]
> or adapted into its package documentation

:+1: Nice idea. This makes a lot more sense in the Dualization docs than in JuMP.

**Comment by odow** [comment]
The JuMP docs should just hint at the "try this out, sometimes it's better." The Dualization docs have a lot more scope to go into the details for users who want that.

**Pull request closed without merging**